### PR TITLE
Mmc 1499 optimizations

### PR DESCRIPTION
**Pull request by ethanmcc**
@narhwaljames @dikee @ovet @mxreppy

### CODE REVIEW

**Comment by mxreppy** [comment]
I'll kind of have to trust you that this works, as it's hard to follow in the abstract.  is this is a breaking upgrade and if so, should the version bump?

**Comment by ethanmcc** [comment]
@mxreppy There will definitely be a version bump. I just want to confirm that it works with my project before making a release.

**Comment by mxreppy** [comment]
:+1:
